### DISCUSSION TITLE
Running Express server as Sidecar

### DISCUSSION THREAD

**Discussion by mcnaveen** [tauri-apps/tauri#5336] [Q&A] [answered]
Things I have done so far
- Created Tauri App (with React)
- Converted Express App as Executable with pkg
- There are two executables. One for Linux and another one for Windows
- I added scope and external binary to the config

Things I'm not sure about:
- I don't know what to paste into the main.rs file. Because I assume it's what helps to start the binary when the app starts.

Since there is no clear documentation of how to do these things. It's really hard and I'm zero in Rust.

**Comment by FabianLars** [comment, marked as answer]
Like i said on Discord, you don't need Rust to start the sidecar. It's basically as the example in the guide shows: https://tauri.app/v1/guides/building/sidecar/#running-on-javascript
But instead of `execute` you use `spawn`:
```ts
import { Command } from '@tauri-apps/api/shell'
// If you use the global tauri object instead of the npm package use `const { Command } = window.__TAURI__.shell` instead
// `bin/my-sidecar` is the value specified on `tauri.conf.json > tauri > bundle > externalBin`
const command = Command.sidecar('bin/my-sidecar')
const child = await command.spawn()
```
To be clear, this javasript code belongs to your Tauri frontend, not the expressjs code.

----

And if you want to do it in rust (this way you can be sure it's only started once) it's how our [repo example](https://github.com/tauri-apps/tauri/blob/dev/examples/sidecar/src-tauri/src/main.rs) shows, but without stdio stuff since you're using a server:
```rs
tauri::Builder::default()
    .setup(|app| {
      let window = app.get_window("main").unwrap();
      tauri::async_runtime::spawn(async move {
        let (mut rx, mut child) = Command::new_sidecar("my-sidecar") // This is only the name of the binary itself, not the complete path like on the javascript side above
          .expect("failed to setup `app` sidecar")
          .spawn()
          .expect("Failed to spawn packaged node");
      });

      Ok(())
    })
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
```

**Reply by mcnaveen** [reply]
Okay. Cool. I pasted the code snippet inside the main.rs file. Since I assume it's recommended way.

What should I give as the name of the binary? Since I have two? Also, do I need to make changes for windows like the folder structure?

> Upon running `yarn tauri dev` the express server is starting only in the dev mode. But after the build isn't not working.

![image](https://user-images.githubusercontent.com/8493007/193726937-d5e8f011-3652-416c-97b6-73ddf6519136.png)

But there are two executables inside the folder. This is what I got building the express server with pkg.

This is my tauri.config.json file (Based on the sidecar docs)

![image](https://user-images.githubusercontent.com/8493007/193722293-1fc64570-ace3-4b8e-a8c4-170f55505296.png)

![image](https://user-images.githubusercontent.com/8493007/193722374-27b7a6e8-8519-4676-bc14-05f4e81349ed.png)

Is this correct? or do I have to make changes? I'm targeting two platforms. Linux and Windows

Thanks

**Reply by FabianLars** [reply]
The folder/file structure should look like this
```
src-tauri/
    binaries/
        app-x86_64-unknown-linux-gnu
        app-x86_64-pc-windows-msvc.exe
```
and in tauri.conf.json you use `"binaries/app"` for externalBin and the scope. Tauri will select the correct binary when bundling.

**Reply by mcnaveen** [reply]
It was so helpful. Finally, I was able to build the final output. The problem is, in the AppImage express server is not starting for some reason. But when I tested the deb file it was working smoothly as expected.

I'm getting this warning during the build process.

```
warning: unused variable: `rx`
  --> src/main.rs:16:18
   |
16 | ...   let (mut rx, mut child) = Command::new_sidecar("app") // This is only the name of the binary itself, not the complete path like on ...
   |                ^^ help: if this is intentional, prefix it with an underscore: `_rx`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `child`
  --> src/main.rs:16:26
   |
16 | ...   let (mut rx, mut child) = Command::new_sidecar("app") // This is only the name of the binary itself, not the complete path like on ...
   |                        ^^^^^ help: if this is intentional, prefix it with an underscore: `_child`

warning: unused variable: `window`
  --> src/main.rs:14:11
   |
14 |       let window = app.get_window("main").unwrap();
   |           ^^^^^^ help: if this is intentional, prefix it with an underscore: `_window`

warning: variable does not need to be mutable
  --> src/main.rs:16:14
   |
16 | ...   let (mut rx, mut child) = Command::new_sidecar("app") // This is only the name of the binary itself, not the complete path like on ...
   |            ----^^
   |            |
   |            help: remove this `mut`
   |
   = note: `#[warn(unused_mut)]` on by default

warning: variable does not need to be mutable
  --> src/main.rs:16:22
   |
16 | ...   let (mut rx, mut child) = Command::new_sidecar("app") // This is only the name of the binary itself, not the complete path like on ...
   |                    ----^^^^^
   |                    |
   |                    help: remove this `mut`

warning: `myapp` (bin "myapp") generated 5 warnings
```

**Reply by FabianLars** [reply]
These warnings are completely irrelevant. If you still wanna get rid of them just change `let (mut rx, mut child) = ` to `let _ = `.

> in the AppImage express server is not starting for some reason

Appimages are super brittle to be honest. You could start it from a command line and see if it prints any errors - but this could get a little hard to triage, not going to lie.
